### PR TITLE
Initial effort

### PR DESCRIPTION
**Pull request by greg0ire**

### CODE REVIEW

File: src/PageOutOfRange.php [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace DigitalBackstage\Searchable;
+
+final class PageOutOfRange implements SearchableException
```
Comment (achevillard) [inline src/PageOutOfRange.php]: Move Exceptions into `DigitalBackstage\Searchable\Exception` namespace

File: src/ResultSet.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace DigitalBackstage\Searchable;
+
+/**
+ * Returned by a Searchable. This is only a subset of all results (the current
+ * page), but should also contain pagination metadata.
+ *
+ * Page numbers start at 1.
+ */
+interface ResultSet extends \Traversable
```
Comment (achevillard) [inline src/ResultSet.php]: interface Paginable ?

File: src/Searchable.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+interface Searchable
+{
+    /**
+     * @throws SortNotSupported if a SortableSearchQuery is passed but
+     * SortableSearchable is not implemented
+     * @throws UnsupportedSearchField if an unssuported search field is used in
+     * the expression.
+     * @throws UnsupportedSortField if an unssuported sort field is used
+     */
```
Comment (achevillard) [inline src/Searchable.php]: @throws PageOutOfRange

File: src/SearchQuery.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace DigitalBackstage\Searchable;
+
+interface SearchQuery
+{
+    /**
+     * An associative array <field> => ['value' => 42, 'type' => 'integer', ...]
+     */
+    public function expression();
```
Comment (sstok) [inline src/SearchQuery.php]: What if you need nested conditions? or that the current condition level needs to be OR-cased.
And custom types, like handling ValueObjects. Etc.

This some more though and standardization 😃 I know it's complex.

Comment (greg0ire) [reply src/SearchQuery.php]: Yeah this is a bit naive at the moment, should probably be a tree. Maybe I just need an `Expression` interface to abstract this away, and a very simple implementation that does not need several levels at first?

**Review by achevillard — APPROVED** [review]

**Pull request merged**